### ISSUE TITLE
geoip fields trigger error

### ISSUE THREAD

**Issue by cvjones17** [elastic/elasticsearch-hadoop#951]
### What kind an issue is this?

- [ X] Bug report. If you’ve found a bug, please provide a code snippet or test to reproduce it below.   
    The easier it is to track down the bug, the faster it is solved. 
- [ ] Feature Request. Start by telling us what problem you’re trying to solve.  
    Often a solution already exists! Don’t send pull requests to implement new features without 
    first getting our support. Sometimes we leave features out on purpose to keep the project small.

### Issue description

I have an ES index with a geoip field. I create a global temporary view wrapping the index, and then try to querying the geoip field. An error, shown below, is produced. I can query all  the other fields, yields an error, as shown below:


### Steps to reproduce

Code:

 Here's the piece of the mapping defining the geoip field:
```
"geoip" : {
            "dynamic" : "true",
            "properties" : {
              "city_name" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "continent_code" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "country_code2" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "country_code3" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "country_name" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "dma_code" : {
                "type" : "long"
              },
              "ip" : {
                "type" : "ip"
              },
              "latitude" : {
                "type" : "float"
              },
              "location" : {
                "type" : "geo_point"
              },
              "longitude" : {
                "type" : "float"
              },
              "postal_code" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "region_code" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "region_name" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              },
              "timezone" : {
                "type" : "text",
                "fields" : {
                  "keyword" : {
                    "type" : "keyword",
                    "ignore_above" : 256
                  }
                }
              }
            }
          }
```
Then in spark SQL, I define a global temporary view wrapping that index:

```
CREATE  GLOBAL TEMPORARY VIEW table_with_geop USING org.elasticsearch.spark.sql OPTIONS (resource 'index-with-geoip');
```
When I describe the view, I see:
```
+------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+--+
|  col_name                             |                                                                                                                                       data_type                                                                                                                                        | comment  |
+------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+--+
| id                                    | string                                                                                                                                                                                                                                                                                 
| geoip                                 | struct<city_name:string,continent_code:string,country_code2:string,country_code3:string,country_name:string,dma_code:bigint,ip:string,latitude:float,location:struct<lat:double,lon:double>,longitude:float,postal_code:string,region_code:string,region_name:string,timezone:string>  | NULL     |
| other_field                           | double                                                                                                                                                                                                                                                                                 | NULL     |
+------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+--+
```


Strack trace:

```
select geoip from table_with_geop limit 1;
```
or
```
select geoip.continent_code from table_with_geop limit 1;
```

```
Error: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 16.0 failed 4 times, most recent failure: Lost task 0.3 in stage 16.0 (TID 38, 172.31.30.96, executor 0): java.lang.IndexOutOfBoundsException: 1
	at scala.collection.convert.Wrappers$JListWrapper.productElement(Wrappers.scala:85)
	at scala.runtime.ScalaRunTime$$anon$1.next(ScalaRunTime.scala:177)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:251)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:241)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:383)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:60)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:57)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:377)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:231)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:225)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:826)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:826)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

``` 

### Version Info

OS:         :     Ubuntu 14.04
JVM         :   1.8.0_121-b13
Hadoop/Spark:  2.1.0
ES-Hadoop   :  org.elasticsearch:elasticsearch-spark-20_2.11:5.0.1
ES          :  5.2.1

**Comment by cvjones17** [comment]
Tried upgrading to ES-Hadoop org.elasticsearch:elasticsearch-spark-20_2.11:5.2.2, and error persists.

**Comment by jbaiera** [comment]
@cvjones17 Are your geo point fields uniform in their source format? The connector attempts to sense the format of the geo field before start up, but if the format of the data is not homogenous then the connector will not function correctly. Are you able to reproduce this with a limited set of records?

**Comment by cvjones17** [comment]
Some of the subfields of geoip can be empty/null.  
I will try to repro with a limited set of records where all the fields are populated.

**Comment by cvjones17** [comment]
I explored the hypothesis that the error occurs become some documents have missing values in the geoip object. I wasn't able to confirm the hypothesis. Actually, I loaded a single document into an index, which has all of the geoip fields non-null, i.e.,


{
  ...
        "geoip": {
                "timezone": "America/New_York",
                "ip": "123.45.678.98",
                "latitude": 40.74,
                "continent_code": "NA",
                "city_name": "Long Island",
                "country_code2": "US",
                "country_name": "United States",
                "dma_code": 501,
                "country_code3": "US",
                "region_name": "New York",
                "location": [
                        -73.9361,
                40.74
                ],
                "postal_code": "11101",
                "longitude": -73.9361,
                "region_code": "NY"
        },
        ...
}

setup a global temporary view from scratch, e.g.,

CREATE  GLOBAL TEMPORARY VIEW test1 USING org.elasticsearch.spark.sql OPTIONS (resource 'test', pushdown 'true');

and saw the same error when I tried to query view test1, i.e.,

select * from global_temp.test1;
Error: org.apache.spark.SparkException: Job aborted due to stage failure: Task 4 in stage 20.0 failed 4 times, most recent failure: Lost task 4.3 in stage 20.0 (TID 201, 172.31.30.96, executor 0): java.lang.IndexOutOfBoundsException: 1
	at scala.collection.convert.Wrappers$JListWrapper.productElement(Wrappers.scala:85)
	at scala.runtime.ScalaRunTime$$anon$1.next(ScalaRunTime.scala:177)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:251)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:241)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:383)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:60)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:57)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:377)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:231)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:225)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:826)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:826)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by jbaiera** [comment]
@cvjones17 Thanks for the clean reproducing steps. I'll see what I can dig up on this.

**Comment by yaoboxu** [comment]
I also meet this errors,how to solve it ??

**Comment by brommer** [comment]
I am currently trying to read a geo point, but I am seeing the same issue. The schema is showing:
```
[info] root
[info]  |-- geometry: struct (nullable = true)
[info]  |    |-- coordinates: struct (nullable = true)
[info]  |    |    |-- lat: double (nullable = true)
[info]  |    |    |-- lon: double (nullable = true)
[info]  |    |-- type: string (nullable = true)
```

But when e.g. executing "show", I am receiving the exception:

```
java.lang.IndexOutOfBoundsException:1
at scala.collection.convert.Wrappers$JListWrapper.productElement(Wrappers.scala:85)
at scala.runtime.ScalaRunTime$$anon$1.next(ScalaRunTime.scala:177)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:251)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
```

**Comment by brommer** [comment]
As requested in the forum post https://discuss.elastic.co/t/elasticsearch-hadoop-issue-reading-geopoint-as-dataset/133865/3 I am providing some additional information how to reproduce the issue.

I am using Elasticsearch Hadoop v. 6.2.3 with Spark 2.3.0 trying to read data from Elasticsearch as a dataset (I think the dataset part is irrelevant since I receive the same issue without converting it to a dataset and just e.g. calling "show" of the dataframe).

I am using Scala and have following case class:
```
case class MyEntity (
  ...
  geometry: Geometry
  ...
)
```

Geometry looks like this:
```
case class Geometry (
  `type`: String,
  coordinates: GeoPoint
)
```

And GeoPoint looks like this:
```
case class GeoPoint (
  lat: Double,
  lon: Double
)
```

The schema of the data looks as follows:
```
[info] root
...
[info] |-- geometry: struct (nullable = true)
[info] | |-- coordinates: struct (nullable = true)
[info] | | |-- lat: double (nullable = true)
[info] | | |-- lon: double (nullable = true)
[info] | |-- type: string (nullable = true)
...
```

I am reading the data as dataset as follows:
```
val myEntities = sqlContext
  .esDF("indexname/doctype")
  .select(
  ...
  $"geometry",
  ...
)
.as[MyEntity]
```

Without the geometry, I can successfully read the data from Elasticsearch, so it is definitely related to that.

**Comment by microcomet** [comment]
Component Version:

``
elasticsearch: 5.6.8
elasticsearch-hadoop:5.6.8
spark: 2.2.0
``

my es index mapping:
``
  "mappings": {
    "date_test": {
      "properties": {
        "birthday": {
          "type": "date",
          "format": "epoch_millis"
        },
        "applyDate": {
          "type": "date"
        },
        "geo": {
          "properties": {
            "gis": {
              "type": "geo_point"
            },
            "address": {
              "type": "text"
            }
          }
        }
      }
    }
  }
``

I want to read es data into spark Dataset<Row>, this is my code:

``
        SparkConf sparkConf = sparkConf();
        sparkConf.remove("es.read.field.exclude");
        sparkConf.set("es.read.field.include", "geo.*");
        sparkConf.set("es.read.field.as.array.include", "geo");
        Dataset<Row> geoDf = EsSparkSQL.esDF(sqlContext);
        geoDf.printSchema();
        geoDf.show(20);
``

and sparkConf() function is : 

``
    static SparkConf sparkConf() {
        SparkConf sparkConf = new SparkConf();
        sparkConf.set("spark.master", "local[*]");
        sparkConf.set("spark.app.name", "TestFw");
        sparkConf.set("spark.default.parallelism", "21");
        sparkConf.set("spark.driver.maxResultSize", "20g");

        sparkConf.set("es.nodes", "192.168.5.25,192.168.5.26,192.168.5.27");
        sparkConf.set("es.port", "9200");
        sparkConf.set("es.resource", INDEX_NAME);
        sparkConf.set("es.read.metadata", "true");
        sparkConf.set("es.read.metadata.field", "esDocId");
        sparkConf.set("es.read.field.exclude", "geo");
        sparkConf.set("es.mapping.date.rich", "false");

        return sparkConf;
    }
``

then i got error like this:

``
Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2043)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2062)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:336)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:2853)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2153)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2153)
	at org.apache.spark.sql.Dataset$$anonfun$55.apply(Dataset.scala:2837)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:65)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:2836)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2153)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2366)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:245)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:644)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:603)
	at com.taiwan3.data.spark_tools.spark.StructTypeTest.main(StructTypeTest.java:119)
Caused by: scala.MatchError: 20.1000, -121.0023 (of class java.lang.String)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:236)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:241)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$ArrayConverter$$anonfun$toCatalystImpl$2.apply(CatalystTypeConverters.scala:164)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$ArrayConverter.toCatalystImpl(CatalystTypeConverters.scala:164)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$ArrayConverter.toCatalystImpl(CatalystTypeConverters.scala:154)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:379)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:61)
	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:58)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(generated.java:84)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:234)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:228)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
``

And from log, i can see es-hadoop infer geo_point format is

``
root
 |-- geo: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- address: string (nullable = true)
 |    |    |-- gis: struct (nullable = true)
 |    |    |    |-- lat: double (nullable = true)
 |    |    |    |-- lon: double (nullable = true)
``

but in es, my data present like this:

``
{
          "geo": [
            {
              "gis": "20.1000, -121.0023",
              "address": "test12312"
            }
          ]
}
``

How can i resolve it?

**Comment by MadDoxXimba** [comment]
I encounter similar problem as mentionned above if sqlcontext is used alongside a dataset containing geopoint...

**Comment by masseyke** [comment]
I'm hoping to fix this bug, but I'm struggling to reproduce it using elasticsearch-spark-30_2.12-7.15.2.jar on spark-3.2.0. Here is what I have tried:
First, I created a mapping with a geoip field:
```
curl -XPUT "http://localhost:9200/geo" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "text": {"type": "keyword"},
      "location": {
        "type": "geo_point"
      }
    }
  }
}'
```
Then I indexed 4 different types of documents:
```
curl -XPUT "http://localhost:9200/geo/_doc/1" -H 'Content-Type: application/json' -d'
{
  "text": "Geopoint as an object",
  "location": { 
    "lat": 41.12,
    "lon": -71.34
  }
}'
curl -XPUT "http://localhost:9200/geo/_doc/2" -H 'Content-Type: application/json' -d'
{
  "text": "Null geopoint as an object"
}'
curl -XPUT "http://localhost:9200/geo/_doc/3" -H 'Content-Type: application/json' -d'
{
  "text": "Null geopoint as an object",
  "location": null
}'
curl -XPUT "http://localhost:9200/geo2/_doc/4" -H 'Content-Type: application/json' -d'
{
  "text": "Geopoint as an array object",
  "location": [
    {
    "lat": 41.12,
    "lon": -71.34
    }
  ]
}'
```
Then in spark shell I tried various ways of accessing the data to make it blow up:
```
scala> import org.apache.spark.sql._
import org.apache.spark.sql._

scala> val sqlContext = new SQLContext(sc)

scala> sqlContext.read.format("es").option("es.read.field.as.array.include","location").load("geo").select("location").show(20)
+-----------------+
|         location|
+-----------------+
|[{41.12, -71.34}]|
|             null|
|           [null]|
|[{41.12, -71.34}]|
+-----------------+

scala> sqlContext.sql("CREATE TEMPORARY view geoip USING org.elasticsearch.spark.sql OPTIONS (resource 'geo', 'es.read.field.as.array.include' 'location')")
res2: org.apache.spark.sql.DataFrame = []

scala> sqlContext.sql("select location from geoip").show
+-----------------+                                                             
|         location|
+-----------------+
|[{41.12, -71.34}]|
|             null|
|           [null]|
|[{41.12, -71.34}]|
+-----------------+
```
Everything unfortunately works as expected. Does anyone see what I am missing, or have exact steps to reproduce this? I'll pull down an older version of spark -- there is a chance that this was a spark bug that has been fixed in spark 3.

**Comment by masseyke** [comment]
I got no error with the above on spark 2.4.4, so it look like there is something I'm missing.

**Comment by masseyke** [comment]
Oh actually I _was_ able to reproduce the java.lang.IndexOutOfBoundsException on spark 2, if I did not use the 'es.read.field.as.array.include' 'location' option. In spark 3 it fails, but with a different error message. So I changed the command in my previous post from:
```
sqlContext.sql("CREATE TEMPORARY view geoip USING org.elasticsearch.spark.sql OPTIONS (resource 'geo', 'es.read.field.as.array.include' 'location')")
```
to
```
sqlContext.sql("CREATE TEMPORARY view geoip USING org.elasticsearch.spark.sql OPTIONS (resource 'geo')")
```
and that got the error. This is not a bug though -- this is just unfortunately the way that es-hadoop works.
There were several reports of a bug in this issue though, and I'm not sure all of them were the same. I'll leave this issue open for a little while in case there is a bug I haven't reproduced yet.
